minidump-processor/src/stackwalker/mips.rs

```diff
@@ -0,0 +1,396 @@
+    let module = modules.module_at_address(callee.instruction)?;
+    let grand_callee_parameter_size = grand_callee.and_then(|f| f.parameter_size).unwrap_or(0);
+    let has_grand_callee = grand_callee.is_some();
+
+    let mut stack_walker = CfiStackWalker {
+        instruction: callee.instruction,
+        has_grand_callee,
+        grand_callee_parameter_size,
+
+        callee_ctx: ctx,
+        callee_validity: valid,
+
+        // Default to forwarding all callee-saved regs verbatim.
+        // The CFI evaluator may clear or overwrite these values.
+        // The stack pointer and instruction pointer are not included.
+        caller_ctx: ctx.clone(),
+        caller_validity: callee_forwarded_regs(valid),
+
+        stack_memory,
+        endianness: scroll::BE,
+    };
+
+    symbol_provider
+        .walk_frame(module, &mut stack_walker)
+        .await?;
+    let caller_pc = stack_walker.caller_ctx.get_register_always(PROGRAM_COUNTER);
+    let caller_sp = stack_walker.caller_ctx.get_register_always(STACK_POINTER);
+
+    trace!(
+        "unwind: cfi evaluation was successful -- caller_pc: {caller_pc:#016x}, caller_sp: {caller_sp:#016x}"
+    );
+
+    // Do absolutely NO validation! Yep! As long as CFI evaluation succeeds
+    // (which does include pc and sp resolving), just blindly assume the
+    // values are correct. I Don't Like This, but it's what breakpad does and
+    // we should start with a baseline of parity.
+
+    let context = MinidumpContext {
+        raw: stack_walker.caller_ctx.into_ctx(),
+        valid: MinidumpContextValidity::Some(stack_walker.caller_validity),
+    };
+    Some(StackFrame::from_context(context, FrameTrust::CallFrameInfo))
+}
+
+fn callee_forwarded_regs(valid: &MinidumpContextValidity) -> HashSet<&'static str> {
+    match valid {
+        MinidumpContextValidity::All => CALLEE_SAVED_REGS.iter().copied().collect(),
+        MinidumpContextValidity::Some(ref which) => CALLEE_SAVED_REGS
+            .iter()
+            .filter(|&reg| which.contains(reg))
+            .copied()
+            .collect(),
+    }
+}
+
+async fn get_caller_by_scan32<P>(
+    ctx: &Mips32Context,
+    callee: &StackFrame,
+    stack_memory: &MinidumpMemory<'_>,
+    modules: &MinidumpModuleList,
+    symbol_provider: &P,
+) -> Option<StackFrame>
+where
+    P: SymbolProvider + Sync,
+{
+    const MAX_STACK_SIZE: u32 = 1024;
+    const MIN_ARGS: u32 = 4;
+    const POINTER_WIDTH: u32 = 4;
+    trace!("unwind: trying scan");
+    // Stack scanning is just walking from the end of the frame until we encounter
+    // a value on the stack that looks like a pointer into some code (it's an address
+    // in a range covered by one of our modules). If we find such an instruction,
+    // we assume it's an pc value that was pushed by the CALL instruction that created
+    // the current frame. The next frame is then assumed to end just before that
+    // pc value.
```

Comment:
this comment is just totally wrong from what I can tell. `ra` is saved in the middle of the frame because all your locals come before it.

Reply:
also all references to CALL should be jal